spec: expose tedge configs to external clients (#4241)
thin-edge/thin-edge.io: docs/src/references/mqtt-api.md

}
```

## Configuration

A service that owns a `tedge_config` setting (`tedge-agent` for core/device settings, each cloud

> mapper for its own cloud's settings) publishes any setting explicitly marked exposable as a retained

Reviewer: *Explicitly marked exposable* is confusing for the users as they have no mean to mark a setting as exposable. This is a internal concept.

```suggestion
mapper for its own cloud's settings) publishes a non-secret subset of its setting as a retained
```

Author: Re-written in 911c34de80d4cfca023c8968c573a52642376e19

`config` message under its own service topic, one message per value:

```text
te/<service-identifier>/config/<key>
```

`<key>` is the setting's `tedge config` key, kept as a single topic segment (not split on its `.`s), so
it maps directly back to a `tedge config get <key>` invocation. A mapper strips its own cloud (and
profile) qualifier from the key, since the service topic already scopes it: `c8y.url` is published as
`url` under `.../service/tedge-mapper-c8y/config/url`.

```text title="Topic (retain=true)"
te/device/main/service/tedge-agent/config/device.id
```

```text title="Payload"
my-device-01
```

An empty payload means the value is unset, or was removed from configuration since the last time the

> owning service started. A setting is only published this way if it was explicitly marked exposable

Reviewer: Are setting updates dynamically published? When the file is updated? When `tedge config set` is used?

What about default values - not explicitly set in `tedge.toml`? What is the effect of `tedge config unset`?

Author: There is no dynamic update. The published config represents the config the service was started with. Any subsequent changes take effect (and published) only on the next restart. Default values are published and unset values are omitted. Since we switched to publishing a single `config` JSON document message instead of a message-per-key, that published document only contains the set values. I have updated the user facing docs to explicitly state these as well in 911c34d.

}
```

## Configuration

Each %%te%% service publishes a subset of its configuration on its own `config` topic,
so that other clients can read those settings without access to the configuration files:

```text
te/<service-topic-id>/config
```

* `te/device/main/service/tedge-agent/config`: the core device settings from `tedge.toml`
* `te/device/main/service/tedge-mapper-<cloud>/config`: cloud connection settings from `mapper.toml`

The whole set is published as a single JSON document, as a retained message.
Each field name is the corresponding `tedge config` key:
the agent publishes `device.id` as `device.id`.

Each value keeps the type the setting has in `tedge.toml`.
A port is published as a number, a flag as a boolean and a template set as an array,

> so a consumer does not have to parse them back out of strings.

Reviewer 2: I'm slightly unsure whether this is helpful or not. To me, this represents exactly how I would expect the API to behave - why would a port be represented as a string? Obviously if someone writes a script to consume the output of `tedge config get`, then they will need to parse the information from a string, but that's because command output is always a string, not anything we're in control of.

I think the examples also help show that this is what's happening wrt types, and in a less abstract manner.

Author: Same reason as https://github.com/thin-edge/thin-edge.io/pull/4241#discussion_r3781659840. I'll clean that up.

Author: Resolved by aae9032d86d5a81ff82b07e303cbf94f644039ba
